### PR TITLE
Reorganize test directory and split CMakeLists.txt

### PR DESCRIPTION
**Pull request by brawner**
As part of the push toward more coverage in rclcpp, which I'll be working toward over the next couple of weeks, I thought this would be a good time to cleanup the organization of the rclcpp testing directory.

This PR moves the CMakeLists.txt test logic into a subdirectory CMakeLists.txt file and puts all the test source files into `test/rclcpp`. This is similar to the test directory organization of `rcl`. 

This will likely force many open PRs to rebase and make appropriate changes to test files.

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11112/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6418/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9068/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11039/)

Signed-off-by: Stephen Brawner <brawner@gmail.com>

### CODE REVIEW

File: rclcpp/CMakeLists.txt [hunk 1]
```diff
@@ -212,442 +212,10 @@ ament_export_dependencies(statistics_msgs)
 ament_export_dependencies(tracetools)
 
 if(BUILD_TESTING)
-  find_package(ament_cmake_gtest REQUIRED)
   find_package(ament_lint_auto REQUIRED)
   ament_lint_auto_find_test_dependencies()
```
Comment (Karsten1987) [inline rclcpp/CMakeLists.txt:216]: why not putting this into the subfolder as well?

**Comment by brawner** [comment]
I think it helps keep the large number of source files from the rest of the test resources separate. Your suggestion was the first thought I had, but I chose to mirror the organization of `rcl` for a bit more consistency.

File: rclcpp/CMakeLists.txt (hunk 1 above)
Comment (brawner) [reply rclcpp/CMakeLists.txt:216]: It mirrors the organization of rcl's CMakeLists.txt. I can see it making sense either way. `ament_lint_auto` tests don't exactly belong to the tests performed in the test directory, so maybe they should be at the top level? On the other hand if they were in the test subdirectory, it just puts all tests in one location. 

https://github.com/ros2/rcl/blob/9e5bcd2965626b04e5ccc626d97b83b73231e913/rcl/CMakeLists.txt#L116-L121

**Comment by brawner** [comment]
This was a large change, running tests on merge to verify
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11130/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6428/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9079/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11051/)

File: rclcpp/CMakeLists.txt (hunk 1 above)
Comment (Karsten1987) [reply rclcpp/CMakeLists.txt:216]: fair enough to mirror the style of RCL. I was mainly curious why not all test related cmake code would go into its own file. I guess semantically, one could argue that linters count as tests as well and could thus land in the subfolder. But I guess that's bike shedding here.

**Comment by jacobperron** [comment]
@brawner I don't think it's been an issue yet, but consider backporting this sooner than later to avoid conflicts with future backports.

**Pull request merged**